app/src/main/kotlin/com/aliucord/manager/ui/viewmodel/InstallViewModel.kt

```diff
@@ -333,61 +315,111 @@ class InstallViewModel(
+            step.duration = time
+            step.status = Status.SUCCESSFUL
 
+            currentStep = step
             return value.value
         } catch (t: Throwable) {
-            val stacktrace = Log.getStackTraceString(t)
-
-            log += "\n$stacktrace"
-            log += "\nFailed to install Aliucord ${installData.discordType.name} during the ${installStep.name} step!"
+            step.status = Status.UNSUCCESSFUL
+            stacktrace = Log.getStackTraceString(t).trim()
 
+            currentCategory = null
+            currentStep = step
             throw t
         }
     }
 
-    private enum class InstallStep(val log: String) {
-        // Shared install steps
-        NONE(""),
-        APP_ICONS("Patching app icons"),
-        MANIFESTS("Patching apk manifests"),
-        DEX("Adding aliu dex into apk"),
-        REPLACE_LIBS("Replacing libraries"),
-        SIGNING("Signing apks"),
-        INSTALLING("Installing apks"),
-
-        // Kotlin exclusive install steps
-        // DL_KT_APK("Downloading Discord apk"),
-        // DL_ALIUHOOK("Downloading Aliuhook library"),
-
-        // AliuRN exclusive install steps
-        DL_BASE_APK("Downloading base apk"),
-        DL_LIBS_APK("Downloading libraries apk"),
-        DL_LOCALES_APKS("Downloading locale apks"),
-        DL_RES_APKS("Downloading resource apk"),
-        DL_HERMES("Downloading patched hermes & c++ runtime library"),
-        DL_ALIUNATIVE("Downloading AliucordNative library"),
+    var currentCategory: StepCategory? by mutableStateOf(StepCategory.APK_DL)
+
+    // Shared steps
+    val appIconPatch by mutableStateOf(Step(
+        "Patching app icons",
+        Status.QUEUED
+    ))
+
+    val manifestPatch by mutableStateOf(Step(
+        "Patching apk manifests",
+        Status.QUEUED
+    ))
+
+    val dexPatch by mutableStateOf(Step(
+        "Adding aliu dex into apk",
+        Status.QUEUED
+    ))
+
+    val libPatch by mutableStateOf(Step(
+        "Replacing libraries",
+        Status.QUEUED
+    ))
+
+    val signApk by mutableStateOf(Step(
+        "Signing apks",
+        Status.QUEUED
+    ))
+
+    val installingApk by mutableStateOf(Step(
+        "Installing apks",
+        Status.QUEUED
+    ))
+
+    // RN steps
+    val baseApkDl by mutableStateOf(Step(
+        "Downloading base apk",
+        Status.QUEUED
+    ))
+
+    val librariesApkDl by mutableStateOf(Step(
+        "Downloading libraries apk",
+        Status.QUEUED
+    ))
+
+    val localeApkDl by mutableStateOf(Step(
+        "Downloading locale apk",
+        Status.QUEUED
+    ))
+
+    val resourceApkDl by mutableStateOf(Step(
+        "Downloading resource apk",
+        Status.QUEUED
+    ))
+
+    val hermesDl by mutableStateOf(Step(
+        "Downloading hermes & c++ runtime library",
+        Status.QUEUED
+    ))
+
+    val aliuNativeDl by mutableStateOf(Step(
+        "Downloading AliucordNative library",
+        Status.QUEUED
+    ))
```

Comment:
Perhaps using a mutable state map here instead?
ex. make an enum of each step and map its values here
would be a lot more pretty